File: problems/binary-trees/binary_trees.c
#include <stdio.h>
#include <stdlib.h>

typedef struct TreeNode {
    struct TreeNode* left;
    struct TreeNode* right;
} TreeNode;

TreeNode* create_tree(int depth) {
    TreeNode* node = (TreeNode*)malloc(sizeof(TreeNode));
    if (depth > 0) {
        node->left = create_tree(depth - 1);
        node->right = create_tree(depth - 1);
    } else {
        node->left = NULL;
        node->right = NULL;
    }
    return node;
}

int compute_checksum(TreeNode* node) {
    if (node) {
        int checksum = 1 + compute_checksum(node->left) + compute_checksum(node->right);
        free(node);
        return checksum;
    }
    return 0;
}

int main(int argc, char* argv[]) {
    int max_depth = 15;
    if (argc > 1) {
        max_depth = atoi(argv[1]);
    }
    
    // Stretch tree
    int stretch_depth = max_depth + 1;
    TreeNode* stretch_tree = create_tree(stretch_depth);
    int stretch_check = compute_checksum(stretch_tree);
    // Note: compute_checksum also frees the nodes in this C implementation
    
    TreeNode* long_lived_tree = create_tree(max_depth);
    
    long total_check = 0;
    
    for (int depth = 4; depth <= max_depth; depth += 2) {
        int iterations = 1 << (max_depth - depth + 4);
        long check = 0;
        
        for (int i = 1; i <= iterations; ++i) {
            TreeNode* t = create_tree(depth);
            check += compute_checksum(t);
        }
        total_check += check;
    }
    
    long long_lived_check = compute_checksum(long_lived_tree);
    
    return total_check % 256;
}

File: problems/binary-trees/binary_trees.rs
use std::env;

struct TreeNode {
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

fn create_tree(depth: i32) -> Option<Box<TreeNode>> {
    if depth > 0 {
        Some(Box::new(TreeNode {
            left: create_tree(depth - 1),
            right: create_tree(depth - 1),
        }))
    } else {
        Some(Box::new(TreeNode {
            left: None,
            right: None,
        }))
    }
}

fn compute_checksum(node: &Option<Box<TreeNode>>) -> i32 {
    match node {
        Some(n) => 1 + compute_checksum(&n.left) + compute_checksum(&n.right),
        None => 0,
    }
}

fn main() {
    let mut max_depth = 15;
    if let Some(arg) = env::args().nth(1) {
        max_depth = arg.parse().unwrap_or(15);
    }
    
    // Stretch tree
    let stretch_depth = max_depth + 1;
    let stretch_tree = create_tree(stretch_depth);
    let _stretch_check = compute_checksum(&stretch_tree);
    drop(stretch_tree);
    
    let long_lived_tree = create_tree(max_depth);
    
    let mut total_check: i64 = 0;
    
    let mut depth = 4;
    while depth <= max_depth {
        let iterations = 1 << (max_depth - depth + 4);
        let mut check: i64 = 0;
        
        for _ in 1..=iterations {
            let t = create_tree(depth);
            check += compute_checksum(&t) as i64;
            // t is dropped here
        }
        
        total_check += check;
        depth += 2;
    }
    
    let _long_lived_check = compute_checksum(&long_lived_tree);
    
    std::process::exit((total_check % 256) as i32);
}
